### PR TITLE
Fixing and refactoring for fun and profit

### PR DESCRIPTION
**Pull request by luukvanderduim**
Fixed stack-overflow stack-sizes on default Rayon thread pool bit us.
Added command line options for:
	- stack size
	- image size
	- samples (clarification needed)
	- number of threads

Fixed out of bounds due to unneeded indexing
Added module for colour and associated types, helpers and implementatiosnj.
Added module for Vec3 and its implementations.
Lots of unneeded 'enhancements' and premature optimizations. Guilty as char.ged.

Likely borked the literate docs!

### CODE REVIEW

**Comment by jhidding** [comment]
That's right you borked the literate docs! I cannot merge in this state. Question:
- why is the stack-size a flag? can't we compute how big it needs to be?
- what's the deal with nanorand?
- do I understand correctly that modules have changed? Never seen this `pub(crate)` declaration before...
- you changed some code style from Haskell-style to more traditional C-style indentation: in the literate code the Haskell style is a bit easier, since you have no trailing comma while each line has near identical syntax. I presume that Rust doesn't mind a trailing comma too many...

**Comment by luukvanderduim** [comment]
>     * why is the stack-size a flag? can't we compute how big it needs to be?

As soon as I understand `radiance()`'s memory use in relation to samples and the dimensions, yes, I suppose we could.
 
>     * what's the deal with nanorand?
I had frail hopes that we could get away with most image sizes with [`std::thread`](https://doc.rust-lang.org/std/thread/#stack-size) default stack size (2 MiB) by reducing the footprint of the random provider. But it does not seem to matter.
I'll revert to rand. 

>     * do I understand correctly that modules have changed? Never seen this `pub(crate)` declaration before...
I am not sure when fine grained visibility was introduced. I tend to think of it as [explicit visibility](https://doc.rust-lang.org/reference/visibility-and-privacy.html#pubin-path-pubcrate-pubsuper-and-pubself). Because I find the module system still a bit challenging at times, I tend to be explicit at the cost of being really silly (such as here).

`pub(crate)` states this item is public within the module tree of this crate, but not outside, which, in binary crates,  I believe is always equal to `pub`. 
When you are writing a library crate and the item is in - or part of - an item in the crates root, then it may be a practical means to say what it is you want. 

I found a pretty cool example of a private item in a public interface to uphold the [invariant of an angle](https://aloso.github.io/2021/03/28/module-system.html#fine-grained-visibility): 

```Rust 
// pub Angle(pub f32) 
pub Angle(f32)

impl Angle {
    pub fn new(value: f32) -> Self {
        Angle(value.rem_euclid(360.0)) 
    }

    pub fn value(&self) -> f32 {
        self.0
    }
}
```
To expand on this example, one could say
``` pub Angle(pub(in crate::tests) f32);```
because you might want to build invalid Angles in for tests, but otherwise private.

In short: I will remove the sillyness.
> 
>     * you changed some code style from Haskell-style to more traditional C-style indentation: in the literate code the Haskell style is a bit easier, since you have no trailing comma while each line has near identical syntax. I presume that Rust doesn't mind a trailing comma too many...

Not exactly sure what you mean here. 
If this is about code formatting and you wish to enforce formatting in Rust crates, you might want to look into [configuring rustfmt](https://rust-lang.github.io/rustfmt/?version=v1.5.0&search=) to tailor users' tooling to better match Haskell type styling.

**Comment by jhidding** [comment]
I merged your changes manually
